sw/device/lib/dif/README.md

```diff
@@ -50,7 +50,144 @@ memory-mapped hardware, and modifying processor CSRs.
+      during the call. This is for the `default:` case of enum switches (as
+      noted below), and for assertion errors (usually where the function has
+      already caused side-effects so `kDif<ip>BadArg` cannot be used).
+    * `kDif<ip>BadArg`, with value 2, to denote that the caller supplied
+      incorrect arguments. This value must only be returned if the function has
+      not caused any side-effects.
+* `dif_<ip>_<operation>_result_t` is an enum representing more specific failure
+  modes. These specific return code enums can be shared between multiple DIFs
+  that fail in the same way. `<operation>` need not correspond to a DIF name if
+  more than one DIF uses these return codes.
+
+  The first three constants in these specific enums must define the following
+  constants:
+  * `kDif<ip><operation>Ok`, with value `kDif<ip>Ok`,
+  * `kDif<ip><operation>Error`, with value `kDif<ip>Error`, and
+  * `kDif<ip><operation>BadArg`, with value `kDif<ip>BadArg`.
+
+  Additional, specific return code constants must all be defined after
+  these three general constants, and may cover more specific forms of the
+  return codes defined above, including more specific reasons arguments are
+  invalid.
+* `dif_<ip>_toggle_t` is an enum representing an enabled or disabled state. This
+  must define exactly two constants, in no particular order:
+  * `kDif<ip>ToggleEnable`, indicating an enabled state.
+  * `kDif<ip>ToggleDisable`, indicating an enabled state.
+
+  This type is intended to be used as a better-named replacement for `bool`,
+  for operations which are setting wither a behavior is enabled or disabled
+  (such as whether an interrupt is maskable). Values of this type should *not* be
+  used as `if`-statement conditions.
+
+  If no function requires this type, it may be omitted.
+
+#### Lifecycle Functions
+The following functions form the basic functionality for initializing and
+handling the lifetime of a handle.
+
+* `result_t dif_<ip>_init(dif_<ip>_params_t params, dif_<ip>_t *handle);`
+  is a function that initializes `handle` in an implementation-defined way, but
+  does not perform any hardware operations. `handle` may point to uninitialized
+  data but must point to initialized data if this function returns `Ok`.
+* `result_t dif_<ip>_configure(const dif_<ip>_t *handle, dif_<ip>_config_t
+  config);` is a function that configures the hardware managed by `handle` with
+  runtime parameters in an implementation-defined way. This function should be
+  "one-off": it should only need to be called once for the lifetime of the handle.
+
+  If there is no meaningful state to configure, this function may be omitted.
+
+#### Interrupt Management
+
+The following types and functions are a standard inteface for peripherals that
+provide a collection of `INTR_ENABLE`, `INTR_STATE`, and `INTR_TEST` registers
+for interrupt management. A DIF library for a peripheral providing such
+registers must provide this interface.
```

Comment:
Correction: It's worth adding a comment that if a peripheral is `no_auto_intr_regs: true` (in the hjson description), then there is no need to follow this API, even if the non-auto-generated registers are still called `_ENABLE`, `_STATE`, `_TEST`. The only peripheral right now like this is `rv_timer`.